src/main/java/reposense/report/RepoCloner.java

```diff
@@ -0,0 +1,162 @@
+            logger.log(Level.SEVERE, "Branch does not exist! Analysis terminated.", e);
+            generateEmptyRepoReport(outputPath, configs[index]);
+            return null;
+        }
+        prevIndex = index;
+        index = (index + 1) % configs.length;
+        return configs[prevIndex];
+    }
+
+    /**
+     * Cleans up after all repos have been cloned and analyzed
+     */
+    public void cleanup() {
+        deleteDirectory(FileUtil.REPOS_ADDRESS);
+    }
+
+    /**
+     * Returns true if current repo is different from the previously cloned repo.
+     */
+    private boolean isPreviousRepoDifferent() {
+        return prevIndex == index || !configs[prevIndex].getLocation().equals(configs[index].getLocation());
+    }
+
+    /**
+     * Spawns a process to clone repo specified in {@code repoConfig}. Does not wait for process to finish executing.
+     * Should only handle a maximum of one spawned process at any time.
+     */
+    private boolean spawnCloneProcess(String outputPath, RepoConfiguration config) throws IOException {
+        assert(crp == null);
+
+        try {
+            FileUtil.deleteDirectory(config.getRepoRoot());
+            logger.info("Cloning in parallel from " + config.getLocation() + "...");
+            Path rootPath = Paths.get(FileUtil.REPOS_ADDRESS, config.getRepoFolderName());
+            Files.createDirectories(rootPath);
+            crp = runCommandAsync(rootPath, "git clone " + addQuote(config.getLocation().toString()));
+        } catch (RuntimeException | IOException e) {
+            logger.log(Level.WARNING,
+                    "Exception met while trying to clone the repo, will skip this repo.", e);
+            generateEmptyRepoReport(outputPath, config);
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Waits for previously spawned clone process to finish executing.
+     * Should only be called after {@code spawnCloneProcess} has been called.
+     */
+    private boolean waitForCloneProcess(String outputPath, RepoConfiguration config) throws IOException {
+        try {
+            crp.waitForProcess();
+            logger.info("Cloning of " + config.getLocation() + " completed!");
+        } catch (RuntimeException | CommandRunnerProcessException e) {
+            crp = null;
+            logger.log(Level.WARNING,
+                    "Exception met while trying to clone the repo, will skip this repo.", e);
+            generateEmptyRepoReport(outputPath, config);
+            return false;
+        }
+        crp = null;
+
+        try {
+            config.updateBranch();
+        } catch (BranchNotFoundException e) {
+            logger.log(Level.SEVERE, "Branch does not exist! Analysis terminated.", e);
+            generateEmptyRepoReport(outputPath, config);
+            return false;
+        }
+        return true;
+    }
+
+    private void generateEmptyRepoReport(String outputPath, RepoConfiguration config) throws IOException {
```

Comment:
Hmmm, we don't really need to know that we need to generate an empty report within `RepoCloner` if we fail to clone, perhaps rename this method to `handleCloningFailed`?